src/lib.rs

```diff
@@ -43,18 +46,83 @@ pub struct UPClientZenoh {
     query_map: Arc<Mutex<HashMap<String, Query>>>,
     // Save the callback for RPC response
     rpc_callback_map: Arc<Mutex<HashMap<String, Arc<UtransportListener>>>>,
+    // Used to identify different callback
     callback_counter: AtomicU64,
+    // Source UUri in RPC
+    uuri: UUri,
 }
 
 impl UPClientZenoh {
+    /// Create `UPClientZenoh` by applying the Zenoh configuration.
+    /// The `UUri` will be generated randomly.
+    ///
+    /// # Arguments
+    ///
+    /// * `config` - Zenoh configuration. You can refer to [here](https://github.com/eclipse-zenoh/zenoh/blob/0.10.1-rc/DEFAULT_CONFIG.json5) for more configuration details.
+    ///
     /// # Errors
-    /// Will return `Err` if unable to create Zenoh session
+    /// Will return `Err` if unable to create `UPClientZenoh`
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # async_std::task::block_on(async {
+    ///     use up_client_zenoh::UPClientZenoh;
+    ///     use zenoh::config::Config;
+    ///     let upclient = UPClientZenoh::new(Config::default()).await.unwrap();
+    /// # });
+    /// ```
     pub async fn new(config: Config) -> Result<UPClientZenoh, UStatus> {
+        let uuri = UUri {
+            entity: Some(UEntity {
+                name: "default.entity".to_string(),
+                id: Some(u32::from(rand::random::<u16>())),
+                version_major: Some(1),
+                version_minor: None,
+                ..Default::default()
+            })
+            .into(),
+            ..Default::default()
+        };
+        UPClientZenoh::new_with_uuri(config, uuri).await
+    }
+
+    /// Create `UPClientZenoh` by applying the Zenoh configuration and self-defined `UUri`.
+    ///
+    /// # Arguments
+    ///
+    /// * `config` - Zenoh configuration. You can refer to [here](https://github.com/eclipse-zenoh/zenoh/blob/0.10.1-rc/DEFAULT_CONFIG.json5) for more configuration details.
+    /// * `uuri` - The `UUri` which is put in source while sending RPC request.
+    ///
+    /// # Errors
+    /// Will return `Err` if unable to create `UPClientZenoh`
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # async_std::task::block_on(async {
+    ///     use up_client_zenoh::UPClientZenoh;
+    ///     use up_rust::{UEntity, UUri};
+    ///     use zenoh::config::Config;
+    ///     let uuri = UUri {
+    ///         entity: Some(UEntity {
+    ///             name: "default.entity".to_string(),
+    ///             id: Some(u32::from(rand::random::<u16>())),
+    ///             version_major: Some(1),
+    ///             version_minor: None,
+    ///             ..Default::default()
+    ///         })
+    ///         .into(),
+    ///         ..Default::default()
+    ///     };
+    ///     let upclient = UPClientZenoh::new_with_uuri(Config::default(), uuri).await.unwrap();
+    /// # });
+    /// ```
+    pub async fn new_with_uuri(config: Config, uuri: UUri) -> Result<UPClientZenoh, UStatus> {
```

Comment:
Hmm, again I'm thinking of the 99% case of people using this library. They're not gonna want to have to think about what UUri to put or remember that they're supposed to put "rpc.response" in the `UResource`.

IMHO this should probably explicitly take `UAuthority`, `UEntity`, and the `version_major` to do the right thing for the end user.

Reply:
On the other hand, I like that you _store_ it as a `UUri`, that makes sense :slightly_smiling_face:

Reply:
In that case it'd be a different `UUri` tho because for an RPC response, the `UResource` should have "rpc.response" [according to the spec](https://github.com/eclipse-uprotocol/up-spec/blob/main/basics/uri.adoc#3-validators):
![image](https://github.com/eclipse-uprotocol/up-client-zenoh-rust/assets/11622119/02d21dae-d989-4e21-83f7-923a12dddaa5)
whereas the Notification would not

Reply:
Yes, I mean we can store the UUri in UPClientZenoh and provide two functions to get the UUri

* `get_response_uuri`: Will get the UUri and replace the UResource with `rpc.response`
* `get_uuri`: Will get the original UUri

When calling `invoke_method`, we should fill the source address with `get_response_uuri`
When calling `send(notification)`, we should fill the source address with `get_uuri`

Do you think it makes sense?

Reply:
> * `get_response_uuri`: Will get the UUri and replace the UResource with `rpc.response`

> When calling `invoke_method`, we should fill the source address with `get_response_uuri`

Sure -- this would happen within the library itself (within `RpcClient::invoke_method()`), no exposure to the end-user, right? So this function could be `pub (crate)`?

> * `get_uuri`: Will get the original UUri

> When calling `send(notification)`, we should fill the source address with `get_uuri`

Under-the-hood, within `send()`?

Or the end-user would call `get_uuri()` and use that in constructing their Notification message?

I'd lean toward the latter since nothing in the spec says we should be rewriting that in a `up-client-foo-bar` when calling `UTransport::send()`.

-------

In any case -- I think the API should still explicitly ask for `UAuthority`, `UEntity`, and `version_major` to make explicit what we're doing.

WDYT?

Reply:
> Sure -- this would happen within the library itself (within RpcClient::invoke_method()), no exposure to the end-user, right? So this function could be pub (crate)?

Not only. We need to use this if we want to call `send(request)` instead of `invoke_method()`
For example:
https://github.com/eclipse-uprotocol/up-client-zenoh-rust/pull/21/files#diff-b2d3f8f8c9e7487b7199b4393c1f706d07023da1bf6ca24d667b0096e88422d2R125

> I'd lean toward the latter since nothing in the spec says we should be rewriting that in a up-client-foo-bar when calling UTransport::send().

Yes, I mean the latter.

> In any case -- I think the API should still explicitly ask for UAuthority, UEntity, and version_major to make explicit what we're doing.

I'm not sure. Do you think users will call `get_uuri()` to construct the notification message? (I believe notification message should include resource) If not, then I agree with you should only have UAuthority and UEntity.

BTW, why do we need to have `version_major` separately? Shouldn't it be the member of `UEntity`?

Reply:
> Not only. We need to use this if we want to call send(request) instead of invoke_method()

Ah, you're right! I missed this case :slightly_smiling_face:

> Yes, I mean the latter.

:+1: 

> I'm not sure. Do you think users will call get_uuri() to construct the notification message? (I believe notification message should include resource) If not, then I agree with you should only have UAuthority and UEntity.

Hmm. I'm not too sure frankly, it's early and we have no users yet :smile: 

But I wanted to highlight that there are a few aspects to this:
1. What should the API be for `new()`?
   * I think it should take UAuthority and UEntity because those are the only stable elements between the Request source UUri and the Notification source UUri
2. How should people be forming the source UUri in a UMessage for Request?
   * I really like the example you showed for this. Every portion of the UUri is fixed, so this makes a lot of sense.
3. How should people be forming UMessage for Notification?
   * The UAuthority and UEntity are fixed, the UResource is not.
   * The "intention-revealing" option almost seems to me to be to have a function like `get_notification_src_uuri(resource: UResource)` which will return a fully-formed UUri with the UAuthority, UEntity, and UResource

WDYT?

> BTW, why do we need to have `version_major` separately? Shouldn't it be the member of `UEntity`?

Ignore, sorry, I was confused haha.

Reply:
> it's early and we have no users yet

I don't have a strong opinion on using UUri or UAuthority+UEntity. I'll use the latter and we can see if users like it or not in the future.

Reply:
Done.